File: r/multi_backtest_runner/strategies/rsi_mean_reversion.R
# RSI Mean Reversion mit Wilder's Smoothing und expliziten Exits.
#
# Jede Strategie-Datei muss definieren:
#   NAME              - character(1)
#   PARAM_GRID        - named list of vectors (alle Kombinationen werden expandiert)
#   generate_signals  - function(df, ...) -> data.table mit Spalte 'Position'
#
# df hat Spalten: Timestamp, Open, High, Low, Close, Volume

NAME <- "RSI_MeanReversion"

PARAM_GRID <- list(
  period   = c(7L, 10L, 14L, 18L, 21L),
  lower    = c(20, 25, 30),
  upper    = c(70, 75, 80),
  exit_mid = c(TRUE)
)

generate_signals <- function(df, period = 14L, lower = 30, upper = 70, exit_mid = TRUE) {
  # TTR::RSI mit maType = "EMA" entspricht Wilder's Smoothing (alpha = 1/n)
  rsi <- TTR::RSI(df$Close, n = period, maType = "EMA")
  
  n <- nrow(df)
  pos <- numeric(n)
  state <- 0
  
  for (i in seq_len(n)) {
    r <- rsi[i]
    if (is.na(r)) { pos[i] <- 0; next }
    
    if (state == 0) {
      if (r < lower)      state <-  1
      else if (r > upper) state <- -1
    } else if (state == 1) {
      if (exit_mid && r >= 50)  state <- 0
      else if (r > upper)       state <- -1
    } else if (state == -1) {
      if (exit_mid && r <= 50)  state <- 0
      else if (r < lower)       state <- 1
    }
    pos[i] <- state
  }
  
  out <- data.table::copy(df)
  out[, RSI := rsi]
  out[, Position := pos]
  out
}

File: python/strategies/rsi_mean_reversion.py
"""RSI Mean Reversion mit Wilder's Smoothing und expliziten Exits."""
import numpy as np
import pandas as pd

NAME = "RSI_MeanReversion"

PARAM_GRID = {
    "period":      [7, 14, 21],
    "lower":       [20, 25, 30],
    "upper":       [70, 75, 80],
    "exit_mid":    [True],   # Exit bei RSI-Crossing der 50-Linie
}


def _rsi_wilder(close: pd.Series, period: int) -> pd.Series:
    delta = close.diff()
    gain = delta.clip(lower=0.0)
    loss = -delta.clip(upper=0.0)
    # Wilder smoothing == EMA mit alpha = 1/period
    avg_gain = gain.ewm(alpha=1 / period, adjust=False, min_periods=period).mean()
    avg_loss = loss.ewm(alpha=1 / period, adjust=False, min_periods=period).mean()
    rs = avg_gain / (avg_loss + 1e-12)
    return 100 - (100 / (1 + rs))


def generate_signals(df: pd.DataFrame, period=14, lower=30, upper=70, exit_mid=True) -> pd.DataFrame:
    df = df.copy()
    df["RSI"] = _rsi_wilder(df["Close"], period)

    pos = np.zeros(len(df))
    state = 0
    rsi = df["RSI"].values
    for i in range(len(df)):
        r = rsi[i]
        if np.isnan(r):
            pos[i] = 0
            continue
        if state == 0:
            if r < lower:
                state = 1
            elif r > upper:
                state = -1
        elif state == 1:
            if exit_mid and r >= 50:
                state = 0
            elif r > upper:
                state = -1
        elif state == -1:
            if exit_mid and r <= 50:
                state = 0
            elif r < lower:
                state = 1
        pos[i] = state

    df["Position"] = pos
    return df